hikari/interactions/modal_interactions.py

```diff
@@ -0,0 +1,338 @@
+# SOFTWARE.
+"""Models and enums used for Discord's Modals interaction flow."""
+
+from __future__ import annotations
+
+__all__: typing.List[str] = [
+    "ModalResponseTypesT",
+    "ModalInteraction",
+    "ModalInteractionTextInput",
+    "ModalInteraction",
+]
+
+import typing
+
+import attr
+
+from hikari import channels
+from hikari import guilds
+from hikari import messages
+from hikari import snowflakes
+from hikari import traits
+from hikari.interactions import base_interactions
+from hikari.internal import attr_extensions
+
+if typing.TYPE_CHECKING:
+    from hikari import users as _users
+    from hikari.api import special_endpoints
+
+ModalResponseTypesT = typing.Literal[
+    base_interactions.ResponseType.MESSAGE_CREATE, 4, base_interactions.ResponseType.DEFERRED_MESSAGE_CREATE, 5
+]
+"""Type-hint of the response types which are valid for a modal interaction.
+
+The following types are valid for this:
+
+* `hikari.interactions.base_interactions.ResponseType.MESSAGE_CREATE`/`4`
+* `hikari.interactions.base_interactions.ResponseType.DEFERRED_MESSAGE_CREATE`/`5`
+"""
+
+
+@attr.define(kw_only=True, weakref_slot=False)
+class ModalInteractionTextInput(messages.PartialComponent):
+    """A text input component in a modal interaction."""
+
+    custom_id: str = attr.field(repr=True)
+    """Developer set custom ID used for identifying interactions with this modal."""
+
+    value: str = attr.field(repr=True)
+    """Value provided for this text input."""
+
+
+class ModalInteractionActionRow(typing.Protocol):
+    """An action row with only partial text inputs.
+
+    Meant purely for use with ModalInteraction.
+    """
+
+    type: typing.Literal[messages.ComponentType.ACTION_ROW]
+    components: typing.List[ModalInteractionTextInput]
+
+
+@attr_extensions.with_copy
+@attr.define(hash=True, kw_only=True, weakref_slot=False)
+class ModalInteraction(base_interactions.MessageResponseMixin[ModalResponseTypesT]):
+    """Represents a modal interaction on Discord."""
+
+    channel_id: snowflakes.Snowflake = attr.field(eq=False, hash=False, repr=True)
+    """ID of the channel this modal interaction event was triggered in."""
+
+    custom_id: str = attr.field(eq=False, hash=False, repr=True)
+    """The custom id of the modal."""
+
+    guild_id: typing.Optional[snowflakes.Snowflake] = attr.field(eq=False, hash=False, repr=True)
+    """ID of the guild this modal interaction event was triggered in.
+
+    This will be `builtins.None` for modal interactions triggered in DMs.
+    """
+
+    guild_locale: typing.Optional[str] = attr.field(eq=False, hash=False, repr=True)
+    """The preferred language of the guild this modal interaction was triggered in.
+
+    This will be `builtins.None` for modal interactions triggered in DMs.
+
+    !!! note
+        This value can usually only be changed if `COMMUNITY` is in `hikari.guilds.Guild.features`
+        for the guild and will otherwise default to `en-US`.
```

Comment:
This doesnt really need to be documented. In this context, we dont care whether it can be changed to

Reply:
This is for consistency with `ComponentInteraction`